src/main/java/net/snowflake/client/core/bind/BindUploader.java

```diff
@@ -183,25 +194,150 @@ public static synchronized BindUploader newInstance(SFSession session, String st
     }
   }
 
+  /**
+   * Create a new BindUploader which will upload to the given stage path Ensure temporary directory
+   * for file writing exists
+   *
+   * @param session the session to use for uploading binds
+   * @param stageDir the stage path to upload to
+   * @return BindUploader instance
+   */
+  public static synchronized BindUploader newInstanceWithoutDirectory(
+      SFSession session, String stageDir) {
+    return new BindUploader(session, stageDir, null);
+  }
+
   /**
    * Upload the bindValues to stage
    *
    * @param bindValues the bind map to upload
    * @throws BindException if the bind map could not be serialized or upload fails
    */
-  public void upload(Map<String, ParameterBindingDTO> bindValues) throws BindException {
+  @Deprecated
+  public void uploadDeprecated(Map<String, ParameterBindingDTO> bindValues) throws BindException {
     if (!closed) {
       serializeBinds(bindValues);
       putBinds();
     }
   }
 
+  /**
+   * Upload bind parameters via streaming. This replaces previous function upload function where
+   * binds were written to a file which was then uploaded with a PUT statement.
+   *
+   * @param bindValues the bind map to upload
+   * @throws BindException
+   * @throws SQLException
+   */
+  public void upload(Map<String, ParameterBindingDTO> bindValues)
+      throws BindException, SQLException {
+    if (!closed) {
+      List<ColumnTypeDataPair> columns = getColumnValues(bindValues);
+      List<byte[]> csvRows = buildRowsAsBytes(columns);
+      int startIndex = 0;
+      int numBytes = 0;
+      int rowNum = 0;
+      int fileCount = 0;
+
+      while (rowNum < csvRows.size()) {
+        // create a list of byte arrays
+        while (numBytes < inputStreamBufferSize && rowNum < csvRows.size()) {
+          numBytes += csvRows.get(rowNum).length;
+          rowNum++;
+        }
+        // concatenate all byte arrays into 1 and put into input stream
+        ByteBuffer bb = ByteBuffer.allocate(numBytes);
+        for (int i = startIndex; i < rowNum; i++) {
+          bb.put(csvRows.get(i));
+        }
+        byte[] finalBytearray = bb.array();
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(finalBytearray);
```

Comment:
think about using java try-with-resource block so that it can automatically close the resource. http://tutorials.jenkov.com/java-exception-handling/try-with-resources.html

Reply:
Sounds good! Changed to that.